mirgecom/artificial_viscosity.py

```diff
@@ -0,0 +1,340 @@
+        The elementwise constant values between 0 and 1 which indicate the smoothness
+        of a given element.
+    """
+    if not isinstance(u, DOFArray):
+        raise ValueError("u argument must be a DOFArray.")
+
+    actx = u.array_context
+
+    @memoize_in(actx, (smoothness_indicator, "smooth_comp_knl"))
+    def indicator_prg():
+        """Compute the smoothness indicator for all elements."""
+        from arraycontext import make_loopy_program
+        from meshmode.transform_metadata import (ConcurrentElementInameTag,
+                                                 ConcurrentDOFInameTag)
+        import loopy as lp
+        t_unit = make_loopy_program([
+            "{[iel]: 0 <= iel < nelements}",
+            "{[idof]: 0 <= idof < ndiscr_nodes_in}",
+            "{[jdof]: 0 <= jdof < ndiscr_nodes_in}",
+            "{[kdof]: 0 <= kdof < ndiscr_nodes_in}"
+            ],
+            """
+                result[iel,idof] = sum(kdof, vec[iel, kdof]               \
+                                             * vec[iel, kdof]             \
+                                             * modes_active_flag[kdof]) / \
+                                   sum(jdof, vec[iel, jdof]               \
+                                             * vec[iel, jdof]             \
+                                             + 1.0e-12 / ndiscr_nodes_in)
+            """,
+            name="smooth_comp",
+        )
+        return lp.tag_inames(t_unit, {"iel": ConcurrentElementInameTag(),
+                                      "idof": ConcurrentDOFInameTag()})
+
+    @keyed_memoize_in(actx, (smoothness_indicator,
+                             "highest_mode"),
+                      lambda grp: grp.discretization_key())
+    def highest_mode(grp):
+        return actx.from_numpy(
+            np.asarray([1 if sum(mode_id) == grp.order
+                        else 0
+                        for mode_id in grp.mode_ids()])
+        )
+
+    # Convert to modal solution representation
+    modal_map = discr.connection_from_dds(DD_VOLUME, DD_VOLUME_MODAL)
+    uhat = modal_map(u)
+
+    # Compute smoothness indicator value
+    indicator = DOFArray(
+        actx,
+        data=tuple(
+            actx.call_loopy(
+                indicator_prg(),
+                vec=uhat[grp.index],
+                modes_active_flag=highest_mode(grp)
+            )["result"]
+            for grp in discr.discr_from_dd("vol").groups
+        )
+    )
+    indicator = actx.np.log10(indicator + 1.0e-12)
+
+    # Compute artificial viscosity percentage based on indicator and set parameters
+    yesnol = actx.np.greater(indicator, (s0 - kappa))
+    yesnou = actx.np.greater(indicator, (s0 + kappa))
+    sin_indicator = actx.np.where(
+        yesnol,
+        0.5 * (1.0 + actx.np.sin(np.pi * (indicator - s0) / (2.0 * kappa))),
+        0.0 * indicator,
+    )
+    indicator = actx.np.where(yesnou, 1.0 + 0.0 * indicator, sin_indicator)
```

Comment:
numpy arrays work with a scalar field... this would avoid an extra dof array by just using `1.0`

Reply:
I'm not sure if that's extended in the array context though worth checking

Reply:
I do not know if it has changed (i doubt it), but I definitely needed each "rhs" args to be arrays of equal size when using this function in the past.

Reply:
heh, whadda ya know. 
https://github.com/illinois-ceesd/mirgecom/blob/532549a5a5ef504507a685ed4b24629e44649e4b/test/test_various.py#L34

Reply:
Went ahead and fixed this in ([732f005](https://github.com/illinois-ceesd/mirgecom/pull/582/commits/732f005b632bd303fb8bbc61669a95193ce1379e))